scripts/internal/prune_size_model.py

```diff
@@ -0,0 +1,310 @@
+
+        next_larger_num_xgrams = cur_target_num_xgrams
+        while left <= right - tolerance:
+            next_threshold = (left + right) / 2
+            modeled_next_num_xgrams = self.GetModeledNextNumXgrams(next_threshold)
+
+            if modeled_next_num_xgrams < cur_target_num_xgrams:
+                right = next_threshold
+            elif modeled_next_num_xgrams > cur_target_num_xgrams:
+                next_larger_num_xgrams = modeled_next_num_xgrams
+                left = next_threshold
+            else: # modeled_next_num_xgrams == cur_target_num_xgrams, this will probably not happan
+                next_larger_num_xgrams = modeled_next_num_xgrams
+                break
+        if left > right - tolerance:
+            # the while loop is not breaked by the else clause,
+            # so we make sure the modeled_next_num_xgrams >= cur_target_num_xgrams
+            next_threshold = left
+
+        if math.fabs(cur_target_num_xgrams - next_larger_num_xgrams) > 0.2 * cur_target_num_xgrams:
+            LogMessage("Warning: final modeled_next_num_xgrams[{0}] is far "
+                       "from intermediate_target_num_xgrams{1}".format(
+                           next_larger_num_xgrams, cur_target_num_xgrams))
+
+        return (next_threshold, next_larger_num_xgrams)
+
+    def GetModeledNextNumXgrams(self, next_threshold):
+        """
+        This function is a model of how we think the next num-xgrams
+        will vary with the chosen next threshold.
+        We predict the next_num_xgrams based on num_xgrams changed of
+        previous iteration and changes of threshold.
+        """
+
+        prev_num_xgrams = self.GetPrevNumXgrams()
+        cur_threshold = self.GetCurThreshold()
+        cur_num_xgrams = self.GetCurNumXgrams()
+
+        # First predict the num-xgrams we think we'll get if we prune again
+        # with the same threshold 'cur_threshold'. The basic heuristic is
+        # that the num-xgrams will decrease by a factor no greater than
+        # 1.5, and no greater than the cube root of the factor by which
+        # the num-xgrams decreased on the previous iteration.
+        assert prev_num_xgrams >= cur_num_xgrams
+        prev_change_factor = (float(cur_num_xgrams) / prev_num_xgrams) ** (1.0/3.0)
+        baseline_decrease_factor = 1.0 / 1.5
+        # choose the factor from these two, that's closest to one.
+        predicted_decrease_factor = max(prev_change_factor, baseline_decrease_factor)
+
+        # the following gives us the predicted num-xgrams if we were
+        # to prune again with the same threshold 'cur_threshold'.
+        predicted_num_xgrams_if_repeat = predicted_decrease_factor * cur_num_xgrams
+        assert next_threshold >= cur_threshold
+
+        predicted_extra_factor = (next_threshold / cur_threshold) ** (self.xgrams_change_power)
+        return predicted_num_xgrams_if_repeat * predicted_extra_factor
+
+    def AdjustModelForOvershoot(self):
```

Comment:
I'm responding here to your post about the thing not converging right because I think it has to do with this.
I notice some issues with the code..

Firstly, you are doing some  things in this function that don't have anything to do with adjusting the model.
You should have a function Backtrack() that will do the actual backtracking.  Adjusting the model is a separate issue.
For backtracking, if the two most recent thresholds were identical, then you should remove the two most recent elements from the history [i.e. backtrack twice].  If the history would then be empty, then the user-specified initial threshold was too high and you can exit with failure.
Secondly, the way you are adjusting the model needs to be changed to ensure convergence.

The model doesn't have just one parameter [the xgrams_change_power], it has three parameters, although I'll suggest to remove one of them.  Your code currently reads like this:
```
   prev_change_factor = (float(cur_num_xgrams) / prev_num_xgrams) ** (1.0/3.0)
  baseline_decrease_factor = 1.0 / 1.5
```
I think it would be a good idea to get rid of baseline_decrease_factor entirely, as a simplification, so that there will be no 'max' expression.  Be sure to rename variables appropriately.
[if we kept the baseline_decrease_factor, we'd have to update it when experiencing overshoot, by increasing it.]
Have a variable
```
    self.prev_change_power = 0.5   
```
which will replace the 1.0 / 3.0 power in your existing code.
And when you adjust the model, replace your code
```
self.xgrams_change_power *= float(cur_modeled_num_xgrams) / cur_num_xgrams
```
with the following:
```
self.xgrams_change_power *= 1.2
self.prev_change_power *= 1.2
if (self.prev_change_power > 1.0)
    self.prev_change_power = 1.0
```

The problem you are experiencing in your comment here
https://github.com/danpovey/pocolm/pull/45#issuecomment-239734909
is that at the point of failure, the xgrams_change_power is not the important part of the model, the important part is the 'predicted_decrease_factor', which is controlled by other parameters, which you are not updating.






Next, the way you are adjusting the model is not very good.
You need to adjust the